calicoctl/calico_ctl/pool.py

```diff
@@ -122,19 +122,56 @@ def ip_pool_add(cidrs, version, ipip, masquerade):
         print "IP in IP not supported for IPv6 pools"
         sys.exit(1)
 
-    # TODO Reject any cidrs that overlap with existing cidrs in the pool
-    pools = []
-    for cidr in cidrs:
+    current_pool_cidrs = [pool.cidr for pool in client.get_ip_pools(version)]
+    new_pools = []
+
+    # Validate pool CIDRs
+    for new_cidr in cidrs:
+        # Check if new pool overlaps with any existing pool
+        overlapping_cidr = _get_overlapping_cidr(new_cidr, current_pool_cidrs)
+        if overlapping_cidr:
+            print "Cannot add IP pool %s - pool conflicts with existing " \
+                  "pool %s" % (new_cidr, overlapping_cidr)
+            sys.exit(1)
+
+        # Check if this new pool overlaps with any other new pool
+        pending_cidrs = [pool.cidr for pool in new_pools]
+        overlapping_cidr = _get_overlapping_cidr(new_cidr, pending_cidrs)
+        if overlapping_cidr:
+            print "Cannot add IP pool %s - pool overlaps with another " \
+                  "new pool %s" % (new_cidr, overlapping_cidr)
+            sys.exit(1)
+
+        # Append cidr to pending list of new pools to add to Calico
         try:
-            pools.append(IPPool(cidr, ipip=ipip, masquerade=masquerade))
+            new_pools.append(IPPool(new_cidr, ipip=ipip, masquerade=masquerade))
         except InvalidBlockSizeError:
             print "An IPv%s pool must have a prefix length of %s or lower." \
                   "\nGiven: %s.\nNo pools added." % \
-                  (version, BLOCK_PREFIXLEN[version], cidr)
+                  (version, BLOCK_PREFIXLEN[version], new_cidr)
             sys.exit(1)
-    for pool in pools:
-        client.add_ip_pool(version, pool)
 
+    # Make client call to add each pool to Calico
+    for new_pool in new_pools:
+        client.add_ip_pool(version, new_pool)
+
+
+def _get_overlapping_cidr(cidr, cidrs):
+    """
+    Check if the given cidr overlaps with any CIDR in the cidrs list.
+
+    :param cidr: CIDR to check overlap for.
+    :param cidrs: List of CIDRs to check for overlap.
+    :return: The first CIDR in cidrs that overlaps with cidr, or None.
+    """
+    cidr_net = IPNetwork(cidr)
+    for other_cidr in cidrs:
+        other_cidr_net = IPNetwork(other_cidr)
+        if ((cidr_net in other_cidr_net or other_cidr_net in cidr_net) and
+            not (cidr_net == other_cidr_net)):
```

Comment:
Why are we ignoring the equal case? I think I may have saw a comment from rob that its because we allow it through as an "update", right? If so, definitely need a comment explaining that, but more importantly, the name and description of this function suddenly has a loophole. It may be best to move the equal check out to the function that calls this. Might need to think about this a bit...